type_decoders/README.md

```diff
@@ -0,0 +1,787 @@
+
+```ts
+const stringDateDecoder =
+  // this regex verifies that a string is of the form `YYYY-MM-DD`
+  isRegex(/^\d{4}[\/\-](0?[1-9]|1[012])[\/\-](0?[1-9]|[12][0-9]|3[01])$/)
+    .map(value => new Date(value));
+
+const result = stringDateDecoder.decode('2000-01-01'); // returns `DecoderSuccess<Date>`
+
+if (result instanceof DecoderSuccess) {
+  const value: Date = result.value;
+}
+```
+
+This decoder will verify that a string is in a `YYYY-MM-DD` format and, if so, convert the string to a date. The return type of the decoder is `Date`.
+
+## Decoder API
+
+### assert()
+
+```ts
+export function assert<R, V>(decoder: Decoder<R, V>): { (value: V): R; (value: Promise<V>): Promise<R> };
+export function assert<R, V>(decoder: PromiseDecoder<R, V>): (value: V | Promise<V>) => Promise<R>;
+```
+
+`assert()` accepts a single decoder as an argument and returns a new function which can be used to decode the same values as provided decoder. On decode success, the validated value is returned directly and on failure the `DecoderError` is thrown (rather than returned).
+
+Example:
+
+```ts
+const validator = assert(isNumber());
+
+const value: number = validator(1);
+
+const value: number = validator('1'); // will throw a `DecoderError`
+```
+
+### isBoolean()
+
+```ts
+interface IBooleanDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isBoolean(options?: IBooleanDecoderOptions): Decoder<boolean, unknown>;
+```
+
+`isBoolean()` can be used to verify that an unknown value is a `boolean`.
+
+### isString()
+
+```ts
+interface IStringDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isString(options?: IStringDecoderOptions): Decoder<string, unknown>;
+```
+
+`isString()` can be used to verify that an unknown value is a `string`.
+
+### isNumber()
+
+```ts
+interface INumberDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isNumber(options?: INumberDecoderOptions): Decoder<number, unknown>;
+```
+
+`isNumber()` can be used to verify that an unknown value is a `number`.
+
+### isInteger()
+
+```ts
+interface IIntegerDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isInteger(options?: IIntegerDecoderOptions): Decoder<number, unknown>;
+```
+
+`isInteger()` can be used to verify that an unknown value is a whole `number`.
+
+### isUndefined()
+
+```ts
+interface IUndefinedDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isUndefined(options?: IUndefinedDecoderOptions): Decoder<undefined, unknown>;
+```
+
+`isUndefined()` can be used to verify that an unknown value is `undefined`. This is a convenience function for `isExactly(undefined)`.
+
+### isNull()
+
+```ts
+interface INullDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isNull(options?: INullDecoderOptions): Decoder<null, unknown>;
+```
+
+`isNull()` can be used to verify that an unknown value is `null`. This is a convenience function for `isExactly(null)`.
+
+### isRegex()
```

Comment:
Should this be called `isMatch` or similar, since it doesn’t check the input to see if it’s a regular expression, but rather matches the passed regular expression against the input?

Reply:
@j-f1 good suggestion